[Digital Ocean] Code cleanup with no functional modifications

srikiz opened the pull request:
- Using interfaces for better unit testing (haven't added yet, but will create that separately)
- Removing cyclic dependency - moved the code from pkg/resources/digitalocean/cloud.go to upup/pkg/fi/cloudup/do/cloud.go.
- No changes to the functionality, but just moving code so it's easier to maintain going forward.

FYI - @timoreimann

johngmyers commented:
Please edit commit history so reviewers don't have to see the back-and-forth. A series of smaller commits is better, making it easier for reviewers to verify that any given refactoring has no functional modification, but the stream should not add an unrelated file and then delete it later.

timoreimann reviewed (approved):
Looks good from a DO perspective. I'll leave the official approval to those that bear the right powers. :)

timoreimann commented:
/unassign

srikiz commented:
/hold

johngmyers commented:
@srikiz I'm concerned this might be a bit too big to review in one go. It's easier to review if code is moved in one commit and then cleaned up in a separate commit.

srikiz commented:
@johngmyers - that sounds like a plan. Once I get the issues resolved, I will close this pull request, and create a new one with 2 commits as you suggested. I'll just move the file in the first commit and keep the rest of the changes in a different commit.

johngmyers commented:
@srikiz alternatively you could force-push to this PR.

srikiz commented:
@johngmyers - Hopefully, I got it right :)
In my first commit, I just moved over the file from `pkg/resources/digitalocean/cloud.go` to `upup/pkg/cloudup/do/cloud.go` - with the same content.
In my second commit, I have modified the file to make it really work. 

All other changes are updating the dependencies to point to the new interface.
Hopefully this helps, please feel free to let me know if you need any additional changes to make it easier for review.

Thanks !

johngmyers commented on upup/pkg/fi/cloudup/do/cloud.go:
```diff
@@ -39,44 +38,64 @@ import (
 
 	dns dnsprovider.Interface
 
 	// RegionName holds the region, renamed to avoid conflict with Region()
```
Should fix comment to match

srikiz replied:
I have updated this now.

johngmyers commented on upup/pkg/fi/cloudup/do/cloud.go.new:
```diff
@@ -0,0 +1,566 @@
+/*
```
This file should not be added then removed. Please edit it out of the commit stream.

srikiz replied:
Sure, Removed it.

johngmyers commented on protokube/pkg/protokube/do_volume.go:
```diff
@@ -128,9 +134,33 @@ func NewDOVolumes() (*DOVolumes, error) {
+	accessToken := os.Getenv("DIGITALOCEAN_ACCESS_TOKEN")
+	if accessToken == "" {
+		return nil, errors.New("DIGITALOCEAN_ACCESS_TOKEN is required")
+	}
```
This looks like a functional modification. I can't see any corresponding deletion.

srikiz replied:
Protokube had a dependency on resources - digital ocean package and was getting a newCloud object by calling the dependency [here](https://github.com/kubernetes/kops/pull/11592/files?file-filters%5B%5D=.go#diff-c8f5df1bde4e360ef9070eb81ce136cc4d3c99220c0dc157300898b002826967L106)
Now, since that dependency is removed, it directly depends on the godo client and creates a new object here.

srikiz replied:
> Unfortunately, the second commit is to big. I cannot easily follow to verify the lack of functional modifications. The fact that functions are being reordered at the same time as they or other things are being modified makes the diffs extremely hard to follow.
> 
> It also looks like the first commit won't compile, but is fixed up in the second commit. I can work around some of that, but it makes reviewing a harder task.

@johngmyers - Apologize, its creating a problem in reviewing this. My first commit is failing since I just moved the same file over to upup/pkg/fi/cloudup/do. But I fixed all issues in my second commit, but placing of the functions didn't match well with the original code.

johngmyers replied:
I can see the corresponding deletion now. Thanks.

johngmyers reviewed (changes requested):
Unfortunately, the second commit is to big. I cannot easily follow to verify the lack of functional modifications. The fact that functions are being reordered at the same time as they or other things are being modified makes the diffs extremely hard to follow.

It also looks like the first commit won't compile, but is fixed up in the second commit. I can work around some of that, but it makes reviewing a harder task.

srikiz commented:
> Unfortunately, the second commit is to big. I cannot easily follow to verify the lack of functional modifications. The fact that functions are being reordered at the same time as they or other things are being modified makes the diffs extremely hard to follow.
> 
> It also looks like the first commit won't compile, but is fixed up in the second commit. I can work around some of that, but it makes reviewing a harder task.

@johngmyers - Apologize, its creating a problem in reviewing this. My first commit is failing since I just moved the same file over to upup/pkg/fi/cloudup/do. But I fixed all issues in my second commit, but placing of the functions didn't match well with the original code.

johngmyers commented:
(I haven't yet had time to review your latest push.)

@srikiz my primary task as reviewer is to verify the claim that there are no functional modifications. Like any human, I have limited short-term memory, so the fewer things I have to track at once the more practical the task.

So the request is to have a series of self-contained commits, so that I can review one commit, then discard most state about that commit and move to the next. As an exception to the expectation that each commit should be self-contained and compile, it is customary for larger changes to automatically generated files to be in a separate commit.

You might find it helpful to look at some of my refactoring PRs, for example #11219 (though that has functional changes mixed in).

srikiz commented:
/retest

srikiz commented:
@johngmyers - thank you for your time on this. My updated commit stream should hopefully make it easier now. I have re-ordered the functions so it is in-line with the changes in the first commit. Please feel free to let me know if you need further modifications. Thanks again, for the help !!

johngmyers commented on upup/pkg/fi/cloudup/do/cloud.go:
```diff
@@ -28,13 +42,526 @@ const TagNameEtcdClusterPrefix = "etcdCluster-"
+// DOCloud exposes all the interfaces required to operate on DigitalOcean resources
+type DOCloud interface {
+	fi.Cloud
+	//Client() *godo.Client
```
nit: this should be removed

srikiz replied:
Sure, updated.

johngmyers commented on upup/pkg/fi/cloudup/do/cloud.go:
```diff
@@ -28,13 +42,526 @@ const TagNameEtcdClusterPrefix = "etcdCluster-"
+	//Client() *godo.Client
+	DropletsService() godo.DropletsService
+	DropletActionService() godo.DropletActionsService
+	VolumeService() godo.StorageService
```
Since this is specific to DO, perhaps it should stick with the DO terminology.

srikiz replied:
I'll probably keep this as-is for now, and rename this to StorageService as a separate PR.

johngmyers commented on upup/pkg/fi/cloudup/do/cloud.go:
```diff
@@ -28,13 +42,526 @@ const TagNameEtcdClusterPrefix = "etcdCluster-"
+
+	dns dnsprovider.Interface
+
+	// region holds the DO region
```
nit: should be a complete sentence, including punctuation:

```suggestion
	// region holds the DO region.
```

srikiz replied:
Sure, updated.

johngmyers commented on upup/pkg/fi/cloudup/do/cloud.go:
```diff
@@ -148,64 +168,69 @@ func (c *Cloud) DeleteInstance(i *cloudinstances.CloudInstance) error {
-func (c *Cloud) DNS() (dnsprovider.Interface, error) {
-	return c.dns, nil
+func (c *doCloudImplementation) DNS() (dnsprovider.Interface, error) {
+	provider := dns.NewProvider(c.Client)
```
Why doesn't this return `c.dns`?

srikiz replied:
Missed that, updated it.

johngmyers commented on upup/pkg/fi/cloudup/do/cloud.go:
```diff
@@ -28,13 +42,526 @@ const TagNameEtcdClusterPrefix = "etcdCluster-"
+
+// FindVPCInfo is not implemented, it's only here to satisfy the fi.Cloud interface
+func (c *doCloudImplementation) FindVPCInfo(id string) (*fi.VPCInfo, error) {
+	return nil, errors.New("Not implemented")
```
Go errors should always start with lowercase letters

```suggestion
	return nil, errors.New("not implemented")
```

srikiz replied:
sure, updated.

johngmyers commented on upup/pkg/fi/cloudup/do/mock_do_cloud.go:
```diff
@@ -0,0 +1,138 @@
+// 	MockLoadBalancers godo.LoadBalancersService
+// 	MockDomain        godo.DomainsService
+// 	MockActions       godo.ActionsService
+// }
```
Remove commented-out code.

srikiz replied:
Sure, removed.

johngmyers commented on upup/pkg/fi/cloudup/do/mock_do_cloud.go:
```diff
@@ -0,0 +1,138 @@
+
+// FindVPCInfo is not implemented, it's only here to satisfy the fi.Cloud interface
+func (c *doCloudMockImplementation) FindVPCInfo(id string) (*fi.VPCInfo, error) {
+	return nil, errors.New("Not implemented")
```
lowercase

srikiz replied:
sure, updated.

johngmyers commented on upup/pkg/fi/cloudup/do/mock_do_cloud.go:
```diff
@@ -0,0 +1,138 @@
+}
+
+func (c *doCloudMockImplementation) DeleteInstance(instance *cloudinstances.CloudInstance) error {
+	return errors.New("Not tested")
```
lowercase

srikiz replied:
Done.

johngmyers reviewed (approved):
These comments are all small enough to leave to follow-up work.

Thank you, that was much easier to review.

johngmyers commented:
/lgtm

srikiz commented:
/hold cancel
